tasks/README.md

```diff
@@ -0,0 +1,244 @@
+
+## Task execution
+
+The task file is copied to the target and then executed on the target by the task runner.
+
+No arguments are passed to the task when it is executed.
+
+The task file in the module does not need execute permissions set.
+
+The location of the task file varies based on the transport used by the task runner. Task authors have no control over this path.
+
+The environment the task is executed in is also controlled by the task runner. Task authors should document any requirements they have.
+
+Task runners will handle execution on Microsoft Windows specially based on the following mapping for file extensions:
+
+- **.ps1**: powershell
+- **.rb**: puppet agent ruby if available
+- **.pp**: puppet apply
+
+Future releases of the task runner may expand this or add configuration options to the task runner around interpreter choice.
+
+## Input Methods
+
+There are a three input methods available to tasks: [stdin](#stdin), [environment](#environment-variables), and [powershell](#powershell). By default tasks without a `.ps1` extension are passed parameters with both the `stdin` and `environment` input methods. Tasks with a `.ps1` extension are passed parameters with the `powershell` input method by default.
+
+In the future we may support other formats and methods for passing params to the task.
+
+### Stdin
+
+The parameters are passed to the task in a JSON object on `stdin`.
+
+In addition to the tasks parameters, the task runner may inject metaparameters prefixed by '_', for example `_noop`.
+
+### Environment Variables
+
+Params are set as environment variables before the task is executed.
+
+Param names will be prefixed with `PT_` to create the environment variables in the form `PT_<param_name>` (i.e. the value of the `foobar` param is set as `PT_foobar`).
+
+String parameters will not be quoted. Numerical parameters and structured objects will use their JSON representations.
+
+#### Example
+
+##### JSON
+```json
+{
+  "a": 1,
+  "b": "a string",
+  "c": [1, 2, "3"],
+  "d": {"x": {"y": [0]}}
+}
+```
+
+##### Environment Variables
+```
+PT_a=1
+PT_b=a string
+PT_c=[1, 2, "3"]
+PT_d={"x": {"y": [0]}}
+```
+
+### Powershell
+
+The `powershell` input method will pass each param as a named argument to the powershell script. This is the default input method for task files ending with a `.ps1` extension.
+
+## Output handling
+
+### Stdout
+
+Task output is consumed on stdout and made available through the task runner API.
+
+If the output cannot be parsed as a JSON object, one is created by the task runner with the captured stdout stored in the  `_output` key: i.e. `{"_output": "the string of output"}`. Otherwise the parsed JSON object is used.
+
+Reserved output fields:
+- **_output**: this will contain the stdout of the task if it returns plaintext output
```

Comment:
A text summary of the job result. If the job does not return a JSON object the contents of stdout will be put in this key.